SPR-12538 AsyncRestTemplate interceptors

jmnarloch opened the pull request:
Hi,

I would like to propose this change that introduces the interceptors functionality to the AsyncRestTemplate.

The implementation mimics the existing one for RestTemplate though is adjusted to handle ListenableFuture instead.

This is an entry point from which I would like to start and open the discussion. If you see some issues with this implementation just let me know. This change does not maintain the backward compatibility between the RestTemplate interceptors and I don't see a way to achieve that goal.

The motivation behind adding this is the fact that is really convenient to write extensions that automatically plug-in into the AsyncRestTemplate. Spring Cloud/Spring Cloud Sleuth already does that for RestTemplate and enables for instance request tracing.

Related issue: https://jira.spring.io/browse/SPR-12538

jmnarloch committed:
SPR-12538 AsyncRestTemplate interceptors

jmnarloch commented on spring-web/src/main/java/org/springframework/http/client/AsyncInterceptingClientHttpRequest.java:
```diff
@@ -0,0 +1,99 @@
+            if(nextInterceptor.hasNext()) {
+                AsyncClientHttpRequestInterceptor interceptor = nextInterceptor.next();
+                ListenableFuture<ClientHttpResponse> future = interceptor.interceptRequest(request, body, this);
+                return new IdentityListenableFutureAdapter<ClientHttpResponse>(future);
```
I've added this wrapper to streamline the execution of the callbacks, though LIFO callback execution could still break the ordering, making the client code handler being invoked prior the interceptors callbacks.

jmnarloch commented:
@jhoeller @rstoyanchev @poutsma may I ask you on your opinion on this topic?

poutsma commented on spring-web/src/main/java/org/springframework/http/client/InterceptingAsyncClientHttpRequest.java:
```diff
@@ -0,0 +1,114 @@
+ * @author Jakub Narloch
+ * @see InterceptingAsyncClientHttpRequestFactory
+ */
+public class InterceptingAsyncClientHttpRequest extends AbstractBufferingAsyncClientHttpRequest {
```
This class does not have to be public, I believe.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/InterceptingAsyncClientHttpRequest.java:
```diff
@@ -0,0 +1,114 @@
+                AsyncClientHttpRequestInterceptor interceptor = nextInterceptor.next();
+                ListenableFuture<ClientHttpResponse> future = interceptor.interceptRequest(request, body, this);
+                return new IdentityListenableFutureAdapter<ClientHttpResponse>(future);
+            } else {
```
Please refer to [Spring Code Style](https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Code-Style) to format your code: `else` goes on a new line!

poutsma commented:
Looks good to me, except for a few minor comments.

jmnarloch commented:
Thanks, I've addressed all the issues. If you would consider merging this PR, please also take a look at https://github.com/spring-projects/spring-framework/pull/915, there are completely independent, but provide complementary functionality, so that you would be able to intercept a request and for instance log the response payload or do rewriting of the request body.

jmnarloch commented:
Any luck with this pull request?

dgomesbr commented:
+1 since it would help spring-cloud/spring-cloud-commons#36

rstoyanchev commented:
When 4.2.4 goes out this week, we'll be creating a 4.2.x branch and the master branch will be open for 4.3 changes. We'll merge this PR as soon as that's the case.

Thanks.

rstoyanchev commented:
This is now merged. Thanks! 

Note there were a couple of tweaks in AsyncRequestExecution to make sure it allows decoration of the request.
